File: 2019/rs/src/bin/day_11.rs
use std::collections::HashMap;
use rug::Integer;
use std::fs;

use aoc_2019::CPU;

fn paint(program: &[Integer], start_white: bool) -> HashMap<(isize, isize), Integer> {
    let mut robot = CPU::new(&program);
    let (mut x, mut y, mut direction) = (0isize, 0isize, 0i32);
    let mut panels = HashMap::new();
    if start_white {
        panels.insert((0, 0), 1.into());
    }
    // Run until robot halts
    loop {
        // Give the robot its next colour
        robot.input.push_back(panels.get(&(x, y)).cloned().unwrap_or_default());
        if robot.run() == 99 {
            break panels;
        }
        // Paint the current tile
        panels.insert((x, y), robot.output.pop_front().unwrap());
        // Turn left or right
        let turn = if robot.output.pop_front().unwrap() == 0 { -1 } else { 1 };
        direction = (direction + turn).rem_euclid(4);
        // Move forward
        match direction {
            0 => y -= 1,
            1 => x += 1,
            2 => y += 1,
            3 => x -= 1,
            _ => unreachable!(),
        }
    }
}

fn main() {
    let text = fs::read_to_string("../input/input_11.txt").unwrap();
    let program: Vec<_> = text.trim().split(',').map(|s| s.parse().unwrap()).collect();
    println!("Part 1: {}", paint(&program, false).len());
    println!("Part 2:");
    let panels = paint(&program, true);
    let x0 = panels.keys().map(|&(x, _)| x).min().unwrap();
    let y0 = panels.keys().map(|&(_, y)| y).min().unwrap();
    let x1 = panels.keys().map(|&(x, _)| x).max().unwrap();
    let y1 = panels.keys().map(|&(_, y)| y).max().unwrap();
    for y in y0..=y1 {
        for x in x0..=x1 {
            match panels.get(&(x, y)) {
                Some(v) if *v == 1 => print!("#"),
                _ => print!(" "),
            }
        }
        println!();
    }
}

File: python/2019/day_11.py
from pathlib import Path

from intcode import CPU


deltas = [
    (0, -1),  # Up
    (1, 0),  # Right
    (0, 1),  # Down
    (-1, 0),  # Left
]


def paint(program: list[int], start_white: bool = False) -> dict[tuple[int, int], int]:
    robot = CPU(program)
    x = y = 0
    direction = 0  # Robot starts facing up
    panels = {}
    if start_white:
        panels[(0, 0)] = 1
    # Run until the robot halts
    while True:
        # Give the robot its next colour
        robot.input.append(panels.get((x, y), 0))
        if robot.run() == 99:
            break
        # Paint the current tile
        panels[(x, y)] = robot.output.popleft()
        # Turn left or right
        turn =  -1 if robot.output.popleft() == 0 else 1
        direction = (direction + turn) % 4
        # Move forward
        dx, dy = deltas[direction]
        x, y = x + dx, y + dy
    return panels


def main():
    text = (Path(__file__).parent / "../../input/2019/input_11.txt").read_text()
    program = [int(x) for x in text.strip().split(",")]
    # Print the number of painted panels
    print("Part 1:", len(paint(program)))
    print("Part 2:")
    panels = paint(program, start_white=True)
    x0, y0 = min(p[0] for p in panels), min(p[1] for p in panels)
    x1, y1 = max(p[0] for p in panels), max(p[1] for p in panels)
    for y in range(y0, y1 + 1):
        for x in range(x0, x1 + 1):
            print("#" if panels.get((x, y), 0) else " ", end="" if x < x1 else "\n")


if __name__ == "__main__":
    main()
